### DISCUSSION TITLE
How to freeze the feature extraction layers of the model for training?

### DISCUSSION THREAD

**Discussion by DavidZSchneider** [ultralytics/ultralytics#3429] [Q&A]
I want to train the model on new data.
I want to keep the pretrained backbone model which do the feature extraction as it was trained by Ultralytics.

What would be the the correct way to freeze those layers before applying the custom data set training?

**Comment by atharvavaidya14** [comment]
This would be the way they mention: [Transfer learning with frozen layers](https://docs.ultralytics.com/yolov5/tutorials/transfer_learning_with_frozen_layers/)

But also check out #3862 since it may not work as intended

**Reply by glenn-jocher** [reply]
@atharvavaidya14 to freeze the feature extraction layers of the YOLOv8 model during training, you can use the `--freeze` argument followed by the number of layers you wish to freeze. This argument allows you to specify which layers of the model should not be updated during the training process.

For example, if you want to freeze the first 10 layers, you would include `--freeze 10` in your training command. This will keep the pretrained weights in these layers intact while the remaining layers are trained on your custom dataset.

Please refer to the Ultralytics documentation for more details on how to use this argument effectively. Keep in mind that freezing layers is a form of transfer learning and can help to speed up training and potentially improve performance when fine-tuning on a dataset similar to the one the model was originally trained on.

If you encounter any issues or unexpected behavior while freezing layers, please check the existing issues on the GitHub repository, as there may be discussions or solutions related to your concern.

**Reply by itsmefifa** [reply]
Hi, This is my command to train YOLOv8 yolo task=detect mode=train model=./runs/detect/train15/weights/best.pt data=dataset.yaml show=True epochs=1000 imgsz=1024 patience=100 freeze=10 but it error and show that 
raise SyntaxError(string + CLI_HELP_MSG) from e SyntaxError: 'freeze' is not a valid YOLO argument.
The question is 
1. What should I do ? Because I want to do both of fine-tuning and transfer learning.
2. How many layer that I should freeze 
Thank you

**Reply by glenn-jocher** [reply]
Hey there! 😊 It looks like there was a small hiccup with the `freeze` argument. In YOLOv8, you should use the transfer learning method by setting the `transfer` argument in your command. To freeze the layers, you can specify which stage to freeze, such as `--transfer freeze` for freezing up to the classifier layers.

Here's a revised example of your command:
```bash
yolo task=detect mode=train model=./runs/detect/train15/weights/best.pt data=dataset.yaml show=True epochs=1000 imgsz=1024 patience=100 transfer=freeze
```
Just make sure to adjust the `transfer` setting according to your needs!

For "How many layers to freeze?", it often depends on your dataset's similarity to the dataset the model was pre-trained on. A good starting point is freezing the backbone and training the rest. You can experiment with different settings to find the best setup for your specific task.

Hope this helps! 🚀

**Reply by itsmefifa** [reply]
Thank you. But it still have an error
raise SyntaxError(string + CLI_HELP_MSG) from e
SyntaxError: 'transfer' is not a valid YOLO argument. Similar arguments are i.e. ['translate=0.1', 'tracker=botsort.yaml', 'train'].

Arguments received: ['yolo', 'task=detect', 'mode=train', 'model=./runs/detect/train15/weights/best.pt', 'data=dataset.yaml', 'show=True', 'epochs=1000', 'imgsz=1024', 'patience=100', 'transfer=freeze'].
I am confused about how to adjust the transfer setting or you mean I have to setting in args.yaml of pre-trained model. And I want to know how many layer does YOLOv8n.pt has, so I will know the layer of backbone.

**Reply by glenn-jocher** [reply]
Hi there! Apologies for the confusion earlier. 🙏 Looks like my previous advice had a slight misstep. The `transfer` flag isn't supported directly via the CLI in YOLOv8. For applying transfer learning and freezing certain layers, you'd indeed do some modifications directly in the code or config files, rather than as CLI arguments.

Regarding modifying for transfer learning, you should adjust your model's YAML file. For YOLOv8, the structure and layer details are contained within the `.yaml` file corresponding to your model (e.g., for YOLOv8n, it might be `yolov8n.yaml`). Unfortunately, specific commands for freezing layers via CLI aren't currently provided out of the box with YOLOv8.

As for the number of layers in `yolov8n.pt`, the detailed architecture can be reviewed in the corresponding `.yaml` configuration file, which outlines the layers and their connections. This will give you insight into the backbone layers you may wish to freeze.

Remember, exploring and tweaking the `.yaml` configuration provides a good pathway for adapting the model to your needs and is an essential skill in working with models like YOLOv8. Keep experimenting and good luck! 🚀
